jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+        arguments (txd, txid, confirmations).
+        Note that callbacks MUST correctly return True if they
+        recognized the transaction and processed it, and False
+        if not. The True return value will be used to remove
+        the callback from the list.
+        Arguments:
+        `callbacks` - a list of functions with signature as above
+        and return type boolean.
+        `txinfo` - either a txid expected for the transaction, if
+        known, or a tuple of the ordered output set, of the form
+        (('script': script), ('value': value), ..). This can be
+        constructed from jmbitcoin.deserialize output, key "outs",
+        using tuple(). See WalletService.transaction_monitor().
+        `cb_type` - must be one of "all", "unconfirmed", "confirmed";
+        the first type will be called back once for every new
+        transaction, the second only once when the number of
+        confirmations is 0, and the third only once when the number
+        of confirmations is > 0.
+        """
+        if cb_type == "all":
+            # note that in this case, txid is ignored.
+            self.callbacks["all"].extend(callbacks)
+        elif cb_type in ["unconfirmed", "confirmed"]:
+            if txinfo not in self.callbacks[cb_type].keys():
+                self.callbacks[cb_type][txinfo] = []
+            self.callbacks[cb_type][txinfo].extend(callbacks)
+        else:
+            assert False, "Invalid argument: " + cb_type
+
+
+    def start_wallet_monitoring(self, syncresult):
+        """ Once the initialization of the service
+        (currently, means: wallet sync) is complete,
+        we start the main monitoring jobs of the
+        wallet service (currently, means: monitoring
+        all new transactions on the blockchain that
+        are recognised as belonging to the Bitcoin
+        Core wallet with the JM wallet's label).
+        """
+        if not syncresult:
+            jlog.error("Failed to sync the bitcoin wallet. Shutting down.")
+            reactor.stop()
+        jlog.info("Starting transaction monitor in walletservice")
+        self.monitor_loop = task.LoopingCall(
+            self.transaction_monitor)
+        self.monitor_loop.start(5.0)
+
+    def transaction_monitor(self):
+        """Keeps track of any changes in the wallet (new transactions).
+        Intended to be run as a twisted task.LoopingCall so that this
+        Service is constantly in near-realtime sync with the blockchain.
+        """
+        txlist = self.bci.list_transactions(100)
+        new_txs = []
+        for x in txlist:
+            # process either (a) a completely new tx or
+            # (b) a tx that reached unconf status but we are still
+            # waiting for conf (active_txids)
+            if x['txid'] in self.active_txids or x['txid'] not in self.old_txs:
+                new_txs.append(x)
+        # reset for next polling event:
+        self.old_txs = [x['txid'] for x in txlist]
```

Comment:
For performance reasons I think this should ideally be a `set`, not a `list`, and could be updated in the loop above.

Reply:
I've decided not to address either of these suggestions of set() to avoid further testing. I don't see any scenario currently where the performance would be noticeable. Of course I don't object to that change, though, if someone wants to PR it.

> and could be updated in the loop above.

I don't believe that's the case, since there is a comparison to the current contents of `self.old_txs` inside the loop.

Reply:
I don't think the comparison would cause any problems if you're adding to the set after checking it, but the change is not important anyway.